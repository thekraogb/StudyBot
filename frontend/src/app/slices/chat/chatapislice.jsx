import { apiSlice } from "../../api/apislice";

export const chatApiSlice = apiSlice.injectEndpoints({
  endpoints: (builder) => ({
    createChat: builder.mutation({
      query: () => ({
        url: "/chat",
        method: "POST",
      }),
    }),
    fetchChats: builder.query({
      query: () => "/chat",
    }),
    deleteChat: builder.mutation({
      query: (chatId) => ({
        url: `/chat/${chatId}`,
        method: "DELETE",
      }),
    }),
  }),
});

export const {
  useCreateChatMutation,
  useFetchChatsQuery,
  useDeleteChatMutation,
} = chatApiSlice;
