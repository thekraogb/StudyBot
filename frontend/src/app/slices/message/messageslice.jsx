import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  messages: JSON.parse(localStorage.getItem("messages")) || [
    {
      message: "Hi! Ask me any question related to STEM.",
      sender: "ChatGPT",
    },
  ],
};

const messageSlice = createSlice({
  name: "messages",
  initialState,
  reducers: {
    addMessage: (state, action) => {
      state.messages.push(action.payload);
      localStorage.setItem("messages", JSON.stringify(state.messages));
    },
    setMessages: (state, action) => {
      state.messages = [
        {
          message: "Hi! Ask me any question related to STEM.",
          sender: "ChatGPT",
        },
        ...action.payload,
      ];
      localStorage.setItem("messages", JSON.stringify(state.messages));
    },
    clearMessages: (state) => {
      state.messages = [
        {
          message: "Hi! Ask me any question related to STEM.",
          sender: "ChatGPT",
        },
      ];
      localStorage.removeItem("messages");
    },
  },
});

export const { addMessage, setMessages, clearMessages } = messageSlice.actions;

export default messageSlice.reducer;
