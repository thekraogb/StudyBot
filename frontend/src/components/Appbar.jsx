import AppBar from "@mui/material/AppBar";
import Box from "@mui/material/Box";
import Toolbar from "@mui/material/Toolbar";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";
import IconButton from "@mui/material/IconButton";
import history from "../assets/history.png";
import logout from "../assets/log-out.png";
import newchat from "../assets/new.png";
import { useState } from "react";

const Appbar = () => {
  return (
    <Box position="sticky" sx={{ top: 0, flexGrow: 1, zIndex: 1000 }}>
      <AppBar
        sx={{
          boxShadow: "none",
          paddingTop: "5px",
          backgroundColor: "#FCFCFC",
        }}
      >
        <Toolbar>
          <Button
            color="inherit"
            sx={{ padding: "7px", margin: 0, minWidth: "auto" }}
          >
            <IconButton sx={{ padding: 0, margin: 0 }}>
              <img
                src={history}
                alt="History"
                style={{ width: 24, height: 24 }}
              />
            </IconButton>
          </Button>

          <Typography
            variant="h6"
            component="div"
            sx={{
              flexGrow: 1,
              marginLeft: "22px",
              fontFamily: "Nunito",
              color: "black",
              fontSize: "2.30rem",
              fontWeight: 550,
            }}
          >
            STUDYBOT
          </Typography>
          <Button
            color="inherit"
            sx={{ padding: "7px", marginRight: "5px", minWidth: "auto" }}
          >
            <IconButton sx={{ p: 0 }}>
              <img
                src={newchat}
                alt="History"
                style={{ width: 24, height: 24 }}
              />
            </IconButton>
          </Button>
          <Button
            color="inherit"
            sx={{ padding: "7px", margin: 0, minWidth: "auto" }}
          >
            <IconButton sx={{ p: 0 }}>
              <img
                src={logout}
                alt="History"
                style={{ width: 24, height: 24 }}
              />
            </IconButton>
          </Button>
        </Toolbar>
      </AppBar>
      <div style={{ height: "1px" }}> </div>
    </Box>
  );
};

export default Appbar;
