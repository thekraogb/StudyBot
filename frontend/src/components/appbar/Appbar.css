.appbar-container {
  display: flex;
}

.appbar {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 10;
  background-color: #fcfcfc;
  width: 100%;
  transition: margin-left 0.4s ease, width 0.4s ease;
}

.toolbar {
  display: flex;
  align-items: center;
  padding-top: 0.5%;
}

.icon-button {
  background: none;
  border: none;
  cursor: pointer;
  margin-right: 15px;
  border-radius: 10px;
  padding: 5px;
}

.icon-button:hover {
  background: #f1f1f1;
}

.history-button {
  margin-left: 15px;
}

.title {
  flex-grow: 1;
  font-family: Nunito, sans-serif;
  font-size: 2.3rem;
  font-weight: 500;
  color: black;
  margin-left: 20px;
  padding-bottom: 0%;
  margin-bottom: 0%;
  margin-top: 0.5%;
}

.name {
  background: none;
  border: 1.4px solid black;
  cursor: pointer;
  border-radius: 50%;
  width: 26px;
  height: 26px;
  text-align: center;
  margin: 0%;
}

.name:hover {
  background: #f1f1f1;
}

.name-container {
  background: none;
  border: none;
  cursor: pointer;
  border-radius: 10px;
  padding: 3px;
  margin-right: 25px;
}

.name-container:hover {
  background: #f1f1f1;
}
