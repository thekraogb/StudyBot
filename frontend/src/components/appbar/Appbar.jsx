import React, { useState } from "react";
import history from "../../assets/history.png";
import newchat from "../../assets/new.png";
import "./Appbar.css";
import Sidebar from "../sidebar/sidebar.jsx";
import { useSidebar } from "../../context/sidebarcontext";
import { useDispatch, useSelector } from "react-redux";
import { setChatId } from "../../app/slices/chat/chatslice";
import { setMessages } from "../../app/slices/message/messageslice";
import UserDropdown from "../dropdown/userdropdown";
import { Tooltip } from "react-tooltip";

const Appbar = () => {
  const panelWidth = 300;
  const { open, toggleSidebar } = useSidebar();
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
  const name = useSelector((state) => state.auth.name);
  const dispatch = useDispatch();

  const firstLetter = name ? name.charAt(0).toUpperCase() : "";

  const handleNewChatClick = () => {
    dispatch(setChatId(""));
    dispatch(setMessages([]));
  };

  const handleDropdownToggle = () => {
    setIsDropdownOpen(!isDropdownOpen);
  };

  return (
    <div className="appbar-container">
      <header
        className={"appbar"}
        style={{
          marginLeft: open ? `${panelWidth}px` : "0",
          width: open ? `calc(100% - ${panelWidth}px)` : "100%",
        }}
      >
        <div className="toolbar">
          {!open && (
            <button
              className="icon-button history-button"
              data-tooltip-id="history"
              data-tooltip-content="Chats history"
              onClick={toggleSidebar}
            >
              <img
                src={history}
                alt="History"
                style={{ width: 24, height: 24 }}
              />
            </button>
          )}
          <Tooltip id="history" />
          <h1 className="title">STUDYBOT</h1>

          <button
            className="icon-button"
            data-tooltip-id="newChat"
            data-tooltip-content="New chat"
            onClick={handleNewChatClick}
          >
            <img
              src={newchat}
              alt="New chat"
              style={{ width: 24, height: 24 }}
            />
          </button>
          <Tooltip id="newChat" />
          <div className="name-container" onClick={handleDropdownToggle}>
            <h2 className="name" style={{ padding: "1px" }}>
              {firstLetter}
            </h2>
            {isDropdownOpen && <UserDropdown />}
          </div>
        </div>
      </header>

      <Sidebar />
    </div>
  );
};

export default Appbar;
