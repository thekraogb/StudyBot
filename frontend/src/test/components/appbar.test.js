import React from "react";
import {
  render,
  screen,
  fireEvent,
  act,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import { store } from "../../app/store";
import Appbar from "../../components/appbar/Appbar";
import { useDispatch } from "react-redux";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { SidebarProvider } from "../../context/sidebarcontext";
import { setChatId } from "../../app/slices/chat/chatslice";
import { setMessages } from "../../app/slices/message/messageslice";

afterEach(() => {
  jest.clearAllMocks();
});

jest.mock("react-redux", () => {
  const dispatch = jest.fn();
  return {
    ...jest.requireActual("react-redux"),
    useDispatch: () => dispatch,
  };
});

jest.mock("../../app/api/apislice", () => ({
  __esModule: true,
  apiSlice: {
    reducer: jest.fn(() => ({})),
    middleware: () => (next) => (action) => next(action),
  },
}));

jest.mock("../../app/slices/chat/chatapislice", () => ({
  __esModule: true,
  chatApiSlice: {
    reducer: jest.fn(() => ({})),
  },
  useCreateChatMutation: jest.fn(() => [jest.fn()]),
  useFetchChatsQuery: jest.fn(() => ({
    data: { data: [] },
    refetch: jest.fn(() => ({
      success: true,
      data: { data: [{ chatId: "", title: "" }] },
    })),
  })),
  useDeleteChatMutation: jest.fn(() => [jest.fn()]),
}));

jest.mock("../../app/slices/message/messageapislice", () => ({
  __esModule: true,
  messageApiSlice: {
    reducer: jest.fn(() => ({})),
  },
  useFetchMessagesQuery: jest.fn((chatId, options) => {
    if (chatId && !options.skip) {
      return { data: { data: [{ sender: "", message: "" }] }, error: null };
    }

    return { data: null, error: true };
  }),
  useCreateMessageMutation: jest.fn(() => [jest.fn()]),
}));

jest.mock("../../app/slices/agent/agentapislice", () => ({
  __esModule: true,
  agentApiSlice: {
    reducer: jest.fn(() => ({})),
  },
  useGetMainAnswerMutation: jest.fn(() => [jest.fn()]),
  useGetCommonQuestionAnswerMutation: jest.fn(() => [jest.fn()]),
  useGetSubtopicExplanationMutation: jest.fn(() => [jest.fn()]),
  useGetQuizFeedbackMutation: jest.fn(() => [jest.fn()]),
  useGetQuizAnswerMutation: jest.fn(() => [jest.fn()]),
  useGetQuizQuestionChoicesMutation: jest.fn(() => [jest.fn()]),
  useGetQuizChoiceFeedbackMutation: jest.fn(() => [jest.fn()]),
}));

describe("Appbar", () => {
    beforeEach(async () => {
      await act(async () => {
        render(
          <Provider store={store}>
            <MemoryRouter>
              <SidebarProvider>
                <Appbar />
              </SidebarProvider>
            </MemoryRouter>
          </Provider>
        );
      });
    });

  test("it renders appbar component", async () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <SidebarProvider>
            <Appbar />
          </SidebarProvider>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("it shows UserDropdown component when clicking on name div", async () => {
    const name = screen.getByRole("button", { name: "User Dropdown" });

    await act(async () => {
      fireEvent.click(name);
    });

    expect(await screen.findByText("Log out")).toBeInTheDocument();
  });

});

describe("handleNewChatClick function", () => {
  beforeEach(async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <MemoryRouter>
            <SidebarProvider>
              <Appbar />
            </SidebarProvider>
          </MemoryRouter>
        </Provider>
      );
    });
  });

  test("it dispatchs setChatId and setMessages and shows a tooltip when clicking on new chat button", async () => {
    const dispatch = useDispatch();

    const newChatButton = screen.getByRole("button", { name: /new chat/i });

    await act(async () => {
      fireEvent.mouseOver(screen.getByRole("button", { name: /new chat/i }));
    });

    expect(await screen.findByText("New chat")).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(newChatButton);
    });

    expect(dispatch).toHaveBeenCalledWith(setChatId(""));
    expect(dispatch).toHaveBeenCalledWith(setMessages([]));
  });
});
