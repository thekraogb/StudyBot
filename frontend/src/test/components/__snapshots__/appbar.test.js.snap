// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appbar it renders appbar component 1`] = `
<div
  className="appbar-container"
>
  <header
    className="appbar"
    style={
      {
        "marginLeft": "0",
        "width": "100%",
      }
    }
  >
    <div
      className="toolbar"
    >
      <button
        className="icon-button history-button"
        data-tooltip-content="Chats history"
        data-tooltip-id="history"
        onClick={[Function]}
      >
        <img
          alt="History"
          src="test-file-stub"
          style={
            {
              "height": 24,
              "width": 24,
            }
          }
        />
      </button>
      <h1
        className="title"
      >
        STUDYBOT
      </h1>
      <button
        className="icon-button"
        data-tooltip-content="New chat"
        data-tooltip-id="newChat"
        onClick={[Function]}
      >
        <img
          alt="New chat"
          src="test-file-stub"
          style={
            {
              "height": 24,
              "width": 24,
            }
          }
        />
      </button>
      <div
        aria-label="User Dropdown"
        className="name-container"
        onClick={[Function]}
        role="button"
      >
        <h3
          className="name"
          style={
            {
              "padding": "1px",
            }
          }
        />
      </div>
    </div>
  </header>
  <div
    className="sidebar "
  >
    <div
      className="chat-list"
    >
      <p
        style={
          {
            "color": "#888",
            "fontSize": "1rem",
            "marginTop": "15%",
            "textAlign": "center",
          }
        }
      >
        No chats yet.
      </p>
    </div>
  </div>
</div>
`;
